core/trino-main/src/main/java/io/trino/operator/output/RleAwarePositionsAppender.java

```diff
@@ -0,0 +1,135 @@
+
+    private final BlockPositionEqual equalOperator;
+    private final PositionsAppender flat;
+
+    @Nullable
+    private Block rleValue;
+
+    // FLAT_STATE_MARKER means flat state, EMPTY_STATE_MARKER means initial empty state, positive means rle state and the current rle position count.
+    private int rlePositionCount = EMPTY_STATE_MARKER;
+
+    public RleAwarePositionsAppender(BlockPositionEqual equalOperator, PositionsAppender flat)
+    {
+        this.flat = requireNonNull(flat, "appender is null");
+        this.equalOperator = equalOperator;
+    }
+
+    @Override
+    public void append(IntArrayList positions, Block source)
+    {
+        switchToFlat();
+        flat.append(positions, source);
+    }
+
+    @Override
+    public void appendRle(RunLengthEncodedBlock source)
+    {
+        if (source.getPositionCount() == 0) {
+            return;
+        }
+
+        if (isEmpty()) {
+            // initial empty state, switch to rle state
+            rleValue = source.getValue();
+            rlePositionCount = source.getPositionCount();
+        }
+        else if (rleValue != null) {
+            // we are in the rle state
+            if (equalOperator.equalNullSafe(rleValue, 0, source.getValue(), 0)) {
+                // the values match. we can just add positions.
+                this.rlePositionCount += source.getPositionCount();
+                return;
+            }
+            // RLE values do not match. switch to flat state
+            switchToFlat();
+            flat.appendRle(source);
+        }
+        else {
+            // flat state
+            flat.appendRle(source);
+        }
+    }
+
+    @Override
+    public Block build()
+    {
+        Block result;
+        if (rleValue != null) {
+            result = new RunLengthEncodedBlock(rleValue, rlePositionCount);
+        }
+        else {
+            result = flat.build();
+        }
+
+        reset();
+        return result;
+    }
+
+    private void reset()
+    {
+        rleValue = null;
+        rlePositionCount = EMPTY_STATE_MARKER;
+    }
+
+    @Override
+    public long getRetainedSizeInBytes()
+    {
+        long retainedRleSize = rleValue != null ? rleValue.getRetainedSizeInBytes() : 0;
+        long flatRetainedSize = flat != null ? flat.getRetainedSizeInBytes() : 0;
+        return INSTANCE_SIZE + retainedRleSize + flatRetainedSize;
+    }
+
+    @Override
+    public long getSizeInBytes()
+    {
+        long rleSize = rleValue != null ? rleValue.getSizeInBytes() : 0;
+        long flatSize = flat != null ? flat.getSizeInBytes() : 0;
+        return rleSize + flatSize;
+    }
+
+    private void switchToFlat()
+    {
+        if (rleValue != null) {
+            // we are in the rle state, flatten all RLE blocks
+            flat.appendRle(new RunLengthEncodedBlock(rleValue, rlePositionCount));
+            rleValue = null;
+        }
+        rlePositionCount = FLAT_STATE_MARKER;
```

Comment:
`FLAT_STATE_MARKER` is not checked anywhere. Just set `rlePositionCount` to `-1` and remove constant.

Move this assignment to `if (rleValue != null) {` if